optimum/habana/diffusers/pipelines/stable_diffusion_xl/pipeline_stable_diffusion_xl.py

```diff
@@ -34,15 +38,629 @@
+            add_text_embeds = torch.cat([negative_pooled_prompt_embeds, add_text_embeds], dim=0)
+            add_time_ids = torch.cat([negative_add_time_ids, add_time_ids], dim=0)
+
+        prompt_embeds = prompt_embeds.to(device)
+        add_text_embeds = add_text_embeds.to(device)
+        add_time_ids = add_time_ids.to(device).repeat(batch_size * num_images_per_prompt, 1)
+
+        if ip_adapter_image is not None:
+            image_embeds = self.prepare_ip_adapter_image_embeds(
+                ip_adapter_image, device, batch_size * num_images_per_prompt
+            )
+
+        # 8. Denoising loop
+        num_warmup_steps = max(len(timesteps) - num_inference_steps * self.scheduler.order, 0)
+
+        # 8.1 Apply denoising_end
+        if (
+            self.denoising_end is not None
+            and isinstance(self.denoising_end, float)
+            and self.denoising_end > 0
+            and self.denoising_end < 1
+        ):
+            discrete_timestep_cutoff = int(
+                round(
+                    self.scheduler.config.num_train_timesteps
+                    - (self.denoising_end * self.scheduler.config.num_train_timesteps)
+                )
+            )
+            num_inference_steps = len(list(filter(lambda ts: ts >= discrete_timestep_cutoff, timesteps)))
+            timesteps = timesteps[:num_inference_steps]
+
+        # 9. Optionally get Guidance Scale Embedding
+        timestep_cond = None
+        if self.unet.config.time_cond_proj_dim is not None:
+            guidance_scale_tensor = torch.tensor(self.guidance_scale - 1).repeat(batch_size * num_images_per_prompt)
+            timestep_cond = self.get_guidance_scale_embedding(
+                guidance_scale_tensor, embedding_dim=self.unet.config.time_cond_proj_dim
+            ).to(device=device, dtype=latents.dtype)
+
+        self._num_timesteps = len(timesteps)
+        with self.progress_bar(total=num_inference_steps) as progress_bar:
+            timesteps = [t.item() for t in timesteps]
+            if self.quantized:
+                for i, t in enumerate(timesteps[0:-2]):
+                    if self.interrupt:
+                        continue
+                    latents = self.run_unet(self.unet, latents, timesteps,
+                                            t ,i ,
+                                            add_text_embeds ,add_time_ids ,prompt_embeds, extra_step_kwargs,
+                                            negative_prompt_embeds, negative_add_time_ids, negative_pooled_prompt_embeds,
+                                            num_warmup_steps, progress_bar, callback,
+                                            callback_steps, ip_adapter_image,
+                                            image_embeds, timestep_cond, callback_on_step_end,
+                                            callback_on_step_end_tensor_inputs)
+                for i, t in enumerate(timesteps[-2:], 18):
```

Comment:
Why hard code 18, do we assume 20 timesteps/inference steps?

Reply:
Yes there will be only 20 steps/inference in case of MLPERF SD XL .

JH...!